components/sup/src/manager/user_config_watcher.rs

```diff
@@ -0,0 +1,251 @@
+mod tests {
+    use super::*;
+
+    use std::fs::{remove_file, File};
+    use std::io::Write;
+    use std::path::PathBuf;
+    use std::str::FromStr;
+    use std::thread;
+    use std::time::{Duration, Instant};
+
+    use tempdir::TempDir;
+
+    #[test]
+    fn no_events_at_first() {
+        let service = TestService::default();
+        let mut ucm = UserConfigWatcher::new();
+        ucm.add(&service);
+
+        assert!(!ucm.have_events_for(&service));
+    }
+
+    #[test]
+    fn events_present_after_adding_config() {
+        let service = TestService::default();
+        let mut ucm = UserConfigWatcher::new();
+        ucm.add(&service);
+
+        File::create(service.path().join(USER_CONFIG_FILE)).expect("creating file");
+
+        assert!(wait_for_events(&ucm, &service));
+    }
+
+    #[test]
+    fn events_present_after_changing_config() {
+        let service = TestService::default();
+        let file_path = service.path().join(USER_CONFIG_FILE);
+        let mut ucm = UserConfigWatcher::new();
+
+        ucm.add(&service);
+        let mut file = File::create(&file_path).expect("creating file");
+
+        ucm.reset_events_for(&service);
+
+        file.write_all(b"42").expect(USER_CONFIG_FILE);
+
+        assert!(wait_for_events(&ucm, &service));
+    }
+
+    #[test]
+    fn events_present_after_removing_config() {
+        let service = TestService::default();
+        let file_path = service.path().join(USER_CONFIG_FILE);
+        let mut ucm = UserConfigWatcher::new();
+
+        ucm.add(&service);
+        File::create(&file_path).expect("creating file");
+
+        // Allow the watcher to notice that a file was created.
+        thread::sleep(Duration::from_millis(100));
+
+        ucm.reset_events_for(&service);
+
+        remove_file(&file_path).expect("removing file");
+
+        assert!(wait_for_events(&ucm, &service));
+    }
+
+    fn wait_for_events<T: Serviceable>(ucm: &UserConfigWatcher, service: &T) -> bool {
+        let start = Instant::now();
+        let timeout = Duration::from_millis(1000);
+
+        while start.elapsed() < timeout {
+            if ucm.have_events_for(service) {
+                return true;
+            }
+
+            thread::sleep(Duration::from_millis(10));
+        }
+
+        false
+    }
+
+    struct TestService {
+        name: String,
+        path: PathBuf,
+        service_group: ServiceGroup,
+    }
+
+    impl Serviceable for TestService {
+        fn name(&self) -> &String {
+            &self.name
+        }
+
+        fn path(&self) -> &PathBuf {
+            &self.path
+        }
+
+        fn service_group(&self) -> &ServiceGroup {
+            &self.service_group
+        }
+    }
+
+    impl Default for TestService {
+        fn default() -> Self {
+            Self {
+                name: String::from("foo"),
+                path: TempDir::new("user-config-watcher")
+                    .expect("creating temp dir")
+                    .into_path(),
```

Comment:
One note about this - when you use the `into_path()` function, you also take over the responsibility of removing the created temporary directory. Which you probably don't do, no?

Reply:
Good point, fixed.